src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionsHandler.cs

```diff
@@ -96,20 +135,60 @@ public CodeActionsHandler(ICodeFixService codeFixService, ICodeRefactoringServic
                 return ImmutableArray<CodeAction>.Empty;
             }
 
-            var text = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);
-
-            var textSpan = ProtocolConversions.RangeToTextSpan(selection, text);
-            var codeFixCollections = await codeFixService.GetFixesAsync(document, textSpan, true, cancellationToken).ConfigureAwait(false);
-            var codeRefactorings = await codeRefactoringService.GetRefactoringsAsync(document, textSpan, cancellationToken).ConfigureAwait(false);
+            var (codeFixCollections, codeRefactorings) = await GetCodeFixesAndRefactoringsAsync(
+                document, codeFixService,
+                codeRefactoringService, selection,
+                cancellationToken).ConfigureAwait(false);
 
             var codeActions = codeFixCollections.SelectMany(c => c.Fixes.Select(f => f.Action)).Concat(
-                                codeRefactorings.SelectMany(r => r.CodeActions.Select(ca => ca.action)));
-
-            // Flatten out the nested codeactions.
-            var nestedCodeActions = codeActions.Where(c => c is CodeAction.CodeActionWithNestedActions nc && nc.IsInlinable).SelectMany(nc => nc.NestedCodeActions);
-            codeActions = codeActions.Where(c => !(c is CodeAction.CodeActionWithNestedActions)).Concat(nestedCodeActions);
+                    codeRefactorings.SelectMany(r => r.CodeActions.Select(ca => ca.action)));
 
             return codeActions;
         }
+
+        internal static async Task<IEnumerable<KeyValuePair<CodeAction, CodeActionKind>>> GetCodeActionsAndKindAsync(
```

Comment:
nit - I think this code would make more sense to me if we just iterate separately through the code fixes and refactorings, convert each list to LSP code action, then concat everything at the end rather than having to deal with the tuple stuff.